opm/autodiff/StandardWells_impl.hpp

```diff
@@ -1248,4 +1248,304 @@ namespace Opm
+               const int perf_start = (i_well->second)[1];
+               const int perf_number = (i_well->second)[2];
+               assert((worst_offending_connection >= 0) && (worst_offending_connection < perf_number));
+
+               const int cell_worst_offending_connection = wells_struct->well_cells[perf_start + worst_offending_connection];
+               list_econ_limited.addClosedConnectionsForWell(well_name, cell_worst_offending_connection);
+               const std::string msg = std::string("Connection ") + std::to_string(worst_offending_connection) + std::string(" for well ")
+                                     + well_name + std::string(" will be closed due to econic limit");
+               OpmLog::info(msg);
+
+               if (last_connection) {
+                   list_econ_limited.addShutWell(well_name);
+                   const std::string msg2 = well_name + std::string(" will be shut due to the last connection closed");
+                   OpmLog::info(msg2);
+               }
+           }
+
+       }
+   }
+
+
+
+
+
+    template <class WellState>
+    bool
+    StandardWells::
+    checkRateEconLimits(const WellEconProductionLimits& econ_production_limits,
+                        const WellState& well_state,
+                        const int well_number) const
+    {
+        const Opm::PhaseUsage& pu = fluid_->phaseUsage();
+        const int np = well_state.numPhases();
+
+        if (econ_production_limits.onMinOilRate()) {
+            assert((*active_)[Oil]);
+            const double oil_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Oil ] ];
+            const double min_oil_rate = econ_production_limits.minOilRate();
+            if (std::abs(oil_rate) < min_oil_rate) {
+                return true;
+            }
+        }
+
+        if (econ_production_limits.onMinGasRate() ) {
+            assert((*active_)[Gas]);
+            const double gas_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Gas ] ];
+            const double min_gas_rate = econ_production_limits.minGasRate();
+            if (std::abs(gas_rate) < min_gas_rate) {
+                return true;
+            }
+        }
+
+        if (econ_production_limits.onMinLiquidRate() ) {
+            assert((*active_)[Oil]);
+            assert((*active_)[Water]);
+            const double oil_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Oil ] ];
+            const double water_rate = well_state.wellRates()[well_number * np + pu.phase_pos[ Water ] ];
+            const double liquid_rate = oil_rate + water_rate;
+            const double min_liquid_rate = econ_production_limits.minLiquidRate();
+            if (std::abs(liquid_rate) < min_liquid_rate) {
+                return true;
+            }
+        }
+
+        if (econ_production_limits.onMinReservoirFluidRate()) {
+            OpmLog::warning("Minimum reservoir fluid production rate limit is not supported yet");
```

Comment:
Then probably for all the similar warning, we should add a tag? I think they are pretty similar in term of the times they will appear in the output.